nep-dapi.mediawiki

```diff
@@ -0,0 +1,586 @@
+</pre>
+
+'''Input arguments'''
+
+None
+
+'''Success return value'''
+
+address: string - Address of the connected account
+
+publicKey: string - Address of the connected account
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_DENIED`
+
+'''Example'''
+
+<pre>
+getAccount()
+.then((account: Account) => {
+  const {
+    address,
+    publicKey,
+  } = account;
+
+  console.log('Provider address: ' + address);
+  console.log('Provider public key: ' + publicKey);
+})
+.catch((err: string) => {
+  switch(err) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+
+
+====getBalance====
+Return balance of a specific asset for the given account.
+
+'''Method Interface'''
+
+<pre>
+function getBalance({ address: string, asset: string }): Promise<{ address: string, asset: string, amount: string }}>
+</pre>
+
+'''Input arguments'''
+
+address: string - The asset symbol the check the balance for the connected account
+
+asset: string - The asset symbol the check the balance for the connected account
+
+
+'''Success return value'''
+
+address: string - Address of the connected account
+
+asset: string - Asset symbol for the requested balance
+
+amount: string - Parsed balance of the requested asset
+
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_DENIED`
+
+'''Example'''
+
+<pre>
+getBalance({
+  address: 'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru',
+  asset: 'NKN'
+})
+.then((result: Result) => {
+  const {
+    address,
+    asset,
+    amount,
+  } = result;
+
+  console.log('Connected address: ' + address);
+  console.log('Asset: ' + asset);
+  console.log('Amount: ' + amount);
+})
+.catch((err: string) => {
+  switch(err) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+
+====send====
+Invoke a transfer of a specified amount of a given asset from the connected account to another account.
+
+'''Method Interface'''
+
+<pre>
+function send({ to: string, asset: string, amount: string, remark?: string }): Promise<txid: string>
+</pre>
+
+'''Input arguments'''
+
+to: string - Address of the receiver of the assets to be sent
+
+asset: string - The symbol of the asset to be sent
+
+amount: string - The parsed amount of the asset to be sent
+
+remark?: string - (Optional) Description of the transaction to be made
+
+
+'''Success return value'''
+
+txid: string - The transaction ID of the send invocation
+
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_REFUSED`|`SEND_ERROR`|`MALFORMED_ADDRESS`|`CANCELED`
+
+'''Example'''
+
+<pre>
+send({
+  to: 'ATaWxfUAiBcQixNPq6TvKHEHPQum9bx79d',
+  asset: 'GAS',
```

Comment:
Couple of things regarding send:
1. It should support multiple assets.
2. Asset should be a hash instead of a symbol.
3. Ability to attach a priority fee.

Reply:
Great idea on adding multiple sends per call and the network fees.

On the request to change `asset` to `hash`, I think this comes down to who should be responsible for mapping asset symbols, which are most commonly found on a user interface to their contract hashes. It may make sense to add `scriptHash` as an optional override, but I am thinking if a user trusts a wallet provider to hold their funds, they are more likely to have the right symbol to hash mappings.

Reply:
Looking into adding multiple send outputs, I'm wondering if this is intended just for sending multiple outputs of NEO and GAS, or also for NEP5 transfers. We currently max out at 2-3 free NEP5 transfers per transaction (depending on the underlying contract implementation), within the 10 GAS. So in most cases, it's not even feasible to send transactions for more than 3-4 transfers in 1 transaction, as each would cost an additional 3-4 GAS with the current fee structure.